### PR TITLE
Fix #881 - Implement new fixed window throttling 

### LINKED ISSUE #881: Introduce a dedicated ThrottleWindow to be used with throttling
### Describe your problem(s)
Currently, throttler uses a custom `timeBatch` window [[1](https://github.com/wso2/product-microgateway/blame/jballerina/components/micro-gateway-core/src/main/ballerina/src/gateway/throttler/time_batch.bal)] and a counter aggregator to throttle requests. However, there are some issues with this approach which can affect throttling behaviour.
- it's not a fixed rate time window
- batch won't start from the 0th-time unit
- emits only when the patch expires
- etc... 

### Describe solution / implementation
The idea is to come up with a custom ThrottleWindow implementation, which addresses the above issues. You can find the above-mentioned implementation & a sample here: [[2](https://gist.github.com/grainier/24bcdc8b8d581e01068fb1008e429c2a)]. Note: this implementation requires Ballerina v1.0.5 or above.

[1] https://github.com/wso2/product-microgateway/blame/jballerina/components/micro-gateway-core/src/main/ballerina/src/gateway/throttler/time_batch.bal
[2] https://gist.github.com/grainier/24bcdc8b8d581e01068fb1008e429c2a

---

### PR DESCRIPTION
**Pull request by Rajith90**
### Purpose
This implements the fixed window throttling without ballerina streaming

### Issues

Fixes #881 

### Automation tests
 - Unit tests added: Yes/No
 - Integration tests added: Yes/No

### Tested environments

Not Tested

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [x] Assigned the project
- [x] Validated respective github issues
- [x] Assigned milestone to the github issue(s)

### CODE REVIEW

File: components/micro-gateway-core/pom.xml [hunk 1]
```diff
@@ -39,6 +39,21 @@
             <artifactId>javax.jms-api</artifactId>
             <version>${jmx.api.version}</version>
         </dependency>
+        <dependency>
+            <groupId>org.ops4j.pax.logging</groupId>
+            <artifactId>pax-logging-api</artifactId>
+            <version>1.10.0</version>
```
Comment (praminda) [inline components/micro-gateway-core/pom.xml]: Inline version declaration. Shall we move this to `dependecyManagement`

File: components/micro-gateway-core/src/main/ballerina/src/gateway/native/throttle.bal [hunk 2]
```diff
@@ -0,0 +1,77 @@
+
```
Comment (praminda) [inline components/micro-gateway-core/src/main/ballerina/src/gateway/native/throttle.bal]:
```suggestion
```

File: components/micro-gateway-core/src/main/ballerina/src/gateway/native/throttle.bal [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    int cleanUpFrequency) = @java:Method {
+    name: "initThrottleDataReceiver",
+    class: "org.wso2.micro.gateway.core.throttle.ThrottleDataReceiver"
+} external;
+
+
+public function jPublishNonThrottledEvent(RequestStreamDTO throttleEvent) = @java:Method {
+    name: "processNonThrottledEvent",
+    class: "org.wso2.micro.gateway.core.throttle.ThrottleDataReceiver"
+} external;
```
Comment (praminda) [inline components/micro-gateway-core/src/main/ballerina/src/gateway/native/throttle.bal]:
```suggestion
} external;

```

File: components/micro-gateway-core/src/main/ballerina/src/gateway/utils/throttle_util.bal [hunk 4]
```diff
@@ -146,13 +147,24 @@ public function removeThrottleData(string key) {
...
 
+public function getPolicyDetails(map<json> deployedPolicies, string policyName) returns (map<json>) {
+    if (stringutils:equalsIgnoreCase(policyName, UNLIMITED_TIER) || policyName.length() == 0) {
+        return { count : -1, unitTime :-1, timeUnit : "min", stopOnQuota : true };
+    }
+    return <map<json>>deployedPolicies.get(policyName);
+}
+
+public function calculateUnitTimeInMillis(int unitTime, string timeUnit) {
+
```
Comment (praminda) [inline components/micro-gateway-core/src/main/ballerina/src/gateway/utils/throttle_util.bal]:
```suggestion
```

File: components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/throttle/DataProcessAgent.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+import org.ballerinalang.jvm.values.MapValue;
+
+/**
+ * This class is responsible for executing data processing logic. This class implements runnable interface and
+ * need to execute using thread pool executor.
+ */
+public class DataProcessAgent implements Runnable {
+
+    private ThrottleCounter throttleCounter;
+
```
Comment (praminda) [inline components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/throttle/DataProcessAgent.java]:
```suggestion
```

File: components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/throttle/ThrottleConfigHolder.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.wso2.micro.gateway.core.throttle;
+
+/**
+ * Static holder single pattern implementation to hold the configurations related local throttle processing,
```
Comment (praminda) [inline components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/throttle/ThrottleConfigHolder.java]: Dot at the end is a must for checkstyles
```suggestion
 * Static holder single pattern implementation to hold the configurations related local throttle processing.
```

File: components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/throttle/ThrottleData.java [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.wso2.micro.gateway.core.throttle;
+
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Class to hold throttle counter data against a specific key
```
Comment (praminda) [inline components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/throttle/ThrottleData.java]:
```suggestion
 * Class to hold throttle counter data against a specific key.
```

File: components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/throttle/ThrottleDataCleanUpTask.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+            log.debug("Throttling Cleanup Task Frequency set to " + throttleFrequency);
+        }
+
+        executor.scheduleAtFixedRate(new CleanupTask(), throttleFrequency, throttleFrequency, TimeUnit.SECONDS);
+
+    }
+
+    /**
+     * Add the throttle data instances to be cleaned into the cleanup array list, which is iterated by the cleanup task
+     * which runs periodically
```
Comment (praminda) [inline components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/throttle/ThrottleDataCleanUpTask.java]:
```suggestion
     * which runs periodically.
```

File: components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/throttle/ThrottleDataReceiver.java [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+        //check for a dto
+        try {
+            DataProcessAgent agent = new DataProcessAgent();
+            agent.setDataReference(throttleEvent);
+            executor.execute(agent);
+        } catch (Exception e) {
+            log.error("Error while processing throttling event", e);
+        }
+    }
+
```
Comment (praminda) [inline components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/throttle/ThrottleDataReceiver.java]:
```suggestion
```

File: components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/throttle/ThrottleDataReceiver.java [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+
```
Comment (praminda) [inline components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/throttle/ThrottleDataReceiver.java]:
```suggestion
```

File: components/micro-gateway-core/src/main/ballerina/src/gateway/handlers/cookie_handler.bal [hunk 11]
```diff
@@ -32,7 +32,8 @@ public type CookieAuthHandler object {
 
             //extract cookies from the incoming request
             string authHead = req.getHeader(COOKIE_HEADER);
-            string[] cookies = split(authHead.trim(), ";");
+            string[] cookies =
+            split(authHead.trim(), ";");
```
Comment (AmaliMatharaarachchi) [inline components/micro-gateway-core/src/main/ballerina/src/gateway/handlers/cookie_handler.bal]: Can make it one line

File: distribution/resources/conf/default-micro-gw.conf.template [hunk 12]
```diff
@@ -191,6 +191,14 @@
 
 # Throttling configurations
 [throttlingConfig]
+  [throttlingConfig.nodeLocal]
```
Comment (AmaliMatharaarachchi) [inline distribution/resources/conf/default-micro-gw.conf.template]: Can we add more documentation comments here with which units (e.g.seconds ..) and etc.

File: components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/throttle/DataProcessAgent.java [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+    private long subscriptionTierCount;
+    private long subscriptionTierUnitTime;
+    private String subscriptionTierTimeUnit;
+    private String resourceKey;
+    private long resourceTierCount;
+    private long resourceTierUnitTime;
+    private String resourceTierTimeUnit;
+    private long timestamp;
+
+
```
Comment (AmaliMatharaarachchi) [inline components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/throttle/DataProcessAgent.java]:
```suggestion

```

**Pull request merged**